imagebuildah/build.go

```diff
@@ -675,42 +693,304 @@ func (b *Executor) Execute(ib *imagebuilder.Builder, node *parser.Node) error {
+				// a COPY or ADD command.
+				filesMatch, err := b.copiedFilesMatch(currNode, history[len(history)-1].Created)
+				if err != nil {
+					return "", errors.Wrapf(err, "error checking if copied files match")
+				}
+				if filesMatch {
+					return image.ID, nil
+				}
+			}
+		}
+	}
+	return "", nil
+}
+
+// getImageHistory returns the history of imageID.
+func (b *Executor) getImageHistory(ctx context.Context, imageID string) ([]v1.History, error) {
+	imageRef, err := is.Transport.ParseStoreReference(b.store, "@"+imageID)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error getting image reference %q", imageID)
+	}
+	ref, err := imageRef.NewImage(ctx, nil)
+	if err != nil {
+		return nil, errors.Wrap(err, "error creating new image from reference")
+	}
+	oci, err := ref.OCIConfig(ctx)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error getting oci config of image %q", imageID)
+	}
+	return oci.History, nil
+}
+
+// getCreatedBy returns the command the image at node will be created by.
+func getCreatedBy(node *parser.Node) string {
+	if node.Value == "run" {
+		return "/bin/sh -c " + node.Original[4:]
+	}
+	return "/bin/sh -c #(nop) " + node.Original
+}
+
+// historyMatches returns true if the history of the image matches the lines
+// in the Dockerfile till the point of build we are at.
+// Used to verify whether a cache of the intermediate image exists and whether
+// to run the build again.
+func historyMatches(children []*parser.Node, history []v1.History) bool {
+	i := len(history) - 1
+	for j := len(children) - 1; j >= 0; j-- {
+		instruction := children[j].Original
+		if children[j].Value == "run" {
+			instruction = instruction[4:]
+		}
+		if !strings.Contains(history[i].CreatedBy, instruction) {
+			return false
+		}
+		i--
+	}
+	return true
+}
+
+// getFilesToCopy goes through node to get all the src files that are copied, added or downloaded.
+// It is possible for the Dockerfile to have src as hom*, which means all files that have hom as a prefix.
+// Another format is hom?.txt, which means all files that have that name format with the ? replaced by another character.
+func (b *Executor) getFilesToCopy(node *parser.Node) ([]string, error) {
+	currNode := node.Next
+	var src []string
+	for currNode.Next != nil {
+		if currNode.Next == nil {
+			break
+		}
+		if strings.HasPrefix(currNode.Value, "http://") || strings.HasPrefix(currNode.Value, "https://") {
+			src = append(src, currNode.Value)
+		} else if currNode.Value[len(currNode.Value)-1] == '*' {
+			files, err := ioutil.ReadDir(b.contextDir)
+			if err != nil {
+				return nil, errors.Wrapf(err, "error reading context directory %q", b.contextDir)
+			}
+			for _, file := range files {
+				if strings.HasPrefix(file.Name(), currNode.Value[:len(currNode.Value)-2]) {
+					src = append(src, filepath.Join(b.contextDir, file.Name()))
+				}
+			}
+		} else if strings.Contains(currNode.Value, "?") {
```

Comment:
You should use regex for these checks.  I think the code would be much cleaner.

Reply:
Will it make much of a difference? Won't I have to have two different regex expressions to check against for the two different conditions I am checking for?

Reply:
Actually the best way to handle this is with filepath.Glob
https://golang.org/pkg/path/filepath/#Glob